File: SalmonTrollingAssistant/Domain/Services/LocationServiceImpl.swift
import Foundation
import CoreLocation
import Combine

class LocationServiceImpl: NSObject, LocationService, CLLocationManagerDelegate {
    private let locationManager = CLLocationManager()
    private let locationSubject = CurrentValueSubject<CLLocation?, Never>(nil)
    private let permissionStatusSubject = CurrentValueSubject<LocationPermissionStatus, Never>(.notDetermined)
    
    // Local storage for saved locations
    private let savedLocationsSubject = CurrentValueSubject<[Location], Never>([])
    private let userDefaults = UserDefaults.standard
    private let savedLocationsKey = "savedLocations"
    private let cachedLocationKey = "cachedLocation"
    private let locationCacheTimestampKey = "locationCacheTimestamp"
    
    // Battery optimization settings
    private var isMoving = false
    private var lastSignificantMovement = Date()
    private let significantMovementThreshold: TimeInterval = 300 // 5 minutes
    
    // Location accuracy settings
    private var currentAccuracyMode: AccuracyMode = .balanced
    
    enum AccuracyMode {
        case highAccuracy
        case balanced
        case lowPower
        
        var desiredAccuracy: CLLocationAccuracy {
            switch self {
            case .highAccuracy:
                return kCLLocationAccuracyBest
            case .balanced:
                return kCLLocationAccuracyNearestTenMeters
            case .lowPower:
                return kCLLocationAccuracyHundredMeters
            }
        }
        
        var distanceFilter: CLLocationDistance {
            switch self {
            case .highAccuracy:
                return 5 // 5 meters
            case .balanced:
                return 20 // 20 meters
            case .lowPower:
                return 100 // 100 meters
            }
        }
        
        var updateInterval: TimeInterval {
            switch self {
            case .highAccuracy:
                return 10 // 10 seconds
            case .balanced:
                return 30 // 30 seconds
            case .lowPower:
                return 120 // 2 minutes
            }
        }
    }
    
    override init() {
        super.init()
        setupLocationManager()
        loadSavedLocations()
        loadCachedLocation()
    }
    
    // MARK: - LocationService Protocol
    
    var currentLocation: AnyPublisher<CLLocation?, Never> {
        return locationSubject.eraseToAnyPublisher()
    }
    
    var permissionStatus: AnyPublisher<LocationPermissionStatus, Never> {
        return permissionStatusSubject.eraseToAnyPublisher()
    }
    
    func requestLocationPermission() {
        locationManager.requestWhenInUseAuthorization()
    }
    
    func getCurrentLocation() -> AnyPublisher<CLLocation?, Error> {
        return Future<CLLocation?, Error> { [weak self] promise in
            guard let self = self else {
                promise(.failure(NSError(domain: "LocationService", code: -1, userInfo: [NSLocalizedDescriptionKey: "Service not available"])))
                return
            }
            
            let status = self.permissionStatusSubject.value
            
            if status.isAuthorized {
                // Check if we have a recent location
                if let location = self.locationSubject.value {
                    promise(.success(location))
                    return
                }
                
                // Check if we have a cached location that's still valid
                if let cachedLocation = self.loadCachedLocation() {
                    promise(.success(cachedLocation))
                    
                    // Still request a fresh location in the background
                    self.requestFreshLocation()
                    return
                }
                
                // Request a fresh location
                self.setAccuracyMode(.highAccuracy)
                self.locationManager.requestLocation()
                
                // Set a timeout for the location request
                DispatchQueue.main.asyncAfter(deadline: .now() + 5) { [weak self] in
                    if let location = self?.locationSubject.value {
                        promise(.success(location))
                    } else {
                        // If we still don't have a location, return nil
                        promise(.success(nil))
                        
                        // Reset accuracy mode
                        self?.setAccuracyMode(.balanced)
                    }
                }
            } else {
                promise(.failure(NSError(domain: "LocationService", code: -2, userInfo: [NSLocalizedDescriptionKey: "Location permission denied"])))
            }
        }.eraseToAnyPublisher()
    }
    
    func searchLocations(query: String) -> AnyPublisher<[Location], Error> {
        return Future<[Location], Error> { promise in
            let geocoder = CLGeocoder()
            geocoder.geocodeAddressString(query) { placemarks, error in
                if let error = error {
                    promise(.failure(error))
                    return
                }
                
                let locations = placemarks?.compactMap { placemark -> Location? in
                    guard let name = placemark.name,
                          let location = placemark.location else {
                        return nil
                    }
                    
                    return Location(
                        id: UUID().uuidString,
                        name: name,
                        latitude: location.coordinate.latitude,
                        longitude: location.coordinate.longitude,
                        isSaved: false,
                        notes: nil
                    )
                } ?? []
                
                promise(.success(locations))
            }
        }.eraseToAnyPublisher()
    }
    
    func getSavedLocations() -> AnyPublisher<[Location], Never> {
        return savedLocationsSubject.eraseToAnyPublisher()
    }
    
    func saveLocation(_ location: Location) -> AnyPublisher<Bool, Error> {
        return Future<Bool, Error> { [weak self] promise in
            guard let self = self else {
                promise(.failure(NSError(domain: "LocationService", code: -1, userInfo: [NSLocalizedDescriptionKey: "Service not available"])))
                return
            }
            
            var locations = self.savedLocationsSubject.value
            
            // Check if location already exists
            if let index = locations.firstIndex(where: { $0.id == location.id }) {
                locations[index] = location
            } else {
                // Create a new location with isSaved = true
                var newLocation = location
                newLocation.isSaved = true
                locations.append(newLocation)
            }
            
            self.savedLocationsSubject.send(locations)
            self.saveToDisk(locations)
            promise(.success(true))
        }.eraseToAnyPublisher()
    }
    
    func deleteLocation(withId id: String) -> AnyPublisher<Bool, Error> {
        return Future<Bool, Error> { [weak self] promise in
            guard let self = self else {
                promise(.failure(NSError(domain: "LocationService", code: -1, userInfo: [NSLocalizedDescriptionKey: "Service not available"])))
                return
            }
            
            var locations = self.savedLocationsSubject.value
            locations.removeAll { $0.id == id }
            self.savedLocationsSubject.send(locations)
            self.saveToDisk(locations)
            promise(.success(true))
        }.eraseToAnyPublisher()
    }
    
    // MARK: - CLLocationManagerDelegate
    
    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        guard let newLocation = locations.last else { return }
        
        // Check if this is a significant movement
        if let previousLocation = locationSubject.value {
            let distance = newLocation.distance(from: previousLocation)
            let timeInterval = newLocation.timestamp.timeIntervalSince(previousLocation.timestamp)
            
            // If moved more than 100 meters in less than 5 minutes, consider as moving
            if distance > 100 && timeInterval < 300 {
                isMoving = true
                lastSignificantMovement = Date()
                
                // If moving, increase accuracy
                if currentAccuracyMode != .highAccuracy {
                    setAccuracyMode(.highAccuracy)
                }
            } else {
                // Check if we haven't moved significantly for a while
                if Date().timeIntervalSince(lastSignificantMovement) > significantMovementThreshold {
                    isMoving = false
                    
                    // If not moving, decrease accuracy to save battery
                    if currentAccuracyMode == .highAccuracy {
                        setAccuracyMode(.balanced)
                    }
                }
            }
        } else {
            // First location update
            lastSignificantMovement = Date()
        }
        
        // Update current location
        locationSubject.send(newLocation)
        
        // Cache the location
        cacheLocation(newLocation)
    }
    
    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        print("Location manager failed with error: \(error.localizedDescription)")
        
        // If we get frequent errors, reduce accuracy to save battery
        if currentAccuracyMode == .highAccuracy {
            setAccuracyMode(.balanced)
        }
    }
    
    func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
        updatePermissionStatus(status)
    }
    
    // MARK: - Private Methods
    
    private func setupLocationManager() {
        locationManager.delegate = self
        setAccuracyMode(.balanced) // Start with balanced accuracy
        
        // Enable significant location changes to wake up the app when the device moves significantly
        locationManager.startMonitoringSignificantLocationChanges()
        
        // Check initial permission status
        let status = CLLocationManager.authorizationStatus()
        updatePermissionStatus(status)
    }
    
    private func setAccuracyMode(_ mode: AccuracyMode) {
        currentAccuracyMode = mode
        locationManager.desiredAccuracy = mode.desiredAccuracy
        locationManager.distanceFilter = mode.distanceFilter
        
        // Adjust update interval based on mode
        if mode == .lowPower {
            locationManager.stopUpdatingLocation()
            locationManager.startUpdatingLocation()
            
            // Schedule to stop updates after a while
            DispatchQueue.main.asyncAfter(deadline: .now() + 30) { [weak self] in
                self?.locationManager.stopUpdatingLocation()
            }
        } else {
            locationManager.stopUpdatingLocation()
            locationManager.startUpdatingLocation()
        }
    }
    
    private func requestFreshLocation() {
        // Temporarily increase accuracy for a fresh location
        let previousMode = currentAccuracyMode
        setAccuracyMode(.highAccuracy)
        
        locationManager.requestLocation()
        
        // Reset to previous mode after a short time
        DispatchQueue.main.asyncAfter(deadline: .now() + 5) { [weak self] in
            self?.setAccuracyMode(previousMode)
        }
    }
    
    private func updatePermissionStatus(_ status: CLAuthorizationStatus) {
        let mappedStatus: LocationPermissionStatus
        
        switch status {
        case .notDetermined:
            mappedStatus = .notDetermined
        case .restricted:
            mappedStatus = .restricted
        case .denied:
            mappedStatus = .denied
        case .authorizedAlways:
            mappedStatus = .authorizedAlways
            locationManager.startUpdatingLocation()
        case .authorizedWhenInUse:
            mappedStatus = .authorizedWhenInUse
            locationManager.startUpdatingLocation()
        @unknown default:
            mappedStatus = .notDetermined
        }
        
        permissionStatusSubject.send(mappedStatus)
    }
    
    private func loadSavedLocations() {
        if let data = userDefaults.data(forKey: savedLocationsKey),
           let locations = try? JSONDecoder().decode([Location].self, from: data) {
            savedLocationsSubject.send(locations)
        }
    }
    
    private func saveToDisk(_ locations: [Location]) {
        if let data = try? JSONEncoder().encode(locations) {
            userDefaults.set(data, forKey: savedLocationsKey)
        }
    }
    
    // MARK: - Location Caching
    
    private func cacheLocation(_ location: CLLocation) {
        // Archive the location
        if let archivedLocation = try? NSKeyedArchiver.archivedData(withRootObject: location, requiringSecureCoding: false) {
            userDefaults.set(archivedLocation, forKey: cachedLocationKey)
            userDefaults.set(Date().timeIntervalSince1970, forKey: locationCacheTimestampKey)
        }
    }
    
    private func loadCachedLocation() -> CLLocation? {
        // Check if we have a cached location
        guard let cachedData = userDefaults.data(forKey: cachedLocationKey),
              let timestamp = userDefaults.object(forKey: locationCacheTimestampKey) as? TimeInterval else {
            return nil
        }
        
        // Check if the cached location is still valid (less than 1 hour old)
        let cacheAge = Date().timeIntervalSince1970 - timestamp
        if cacheAge > 3600 { // 1 hour
            return nil
        }
        
        // Unarchive the location
        if let location = try? NSKeyedUnarchiver.unarchiveTopLevelObjectWithData(cachedData) as? CLLocation {
            return location
        }
        
        return nil
    }
}

File: android/SalmonTrollingAssistant/app/src/main/java/com/example/salmontrollingassistant/data/service/LocationServiceImpl.kt
package com.example.salmontrollingassistant.data.service

import android.annotation.SuppressLint
import android.content.Context
import android.content.SharedPreferences
import android.location.Address
import android.location.Geocoder
import android.location.Location
import android.os.Build
import android.os.Looper
import android.os.SystemClock
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import androidx.room.Room
import com.example.salmontrollingassistant.data.db.AppDatabase
import com.example.salmontrollingassistant.data.db.LocationDao
import com.example.salmontrollingassistant.data.db.LocationEntity
import com.example.salmontrollingassistant.domain.model.Location as DomainLocation
import com.example.salmontrollingassistant.domain.service.LocationPermissionStatus
import com.example.salmontrollingassistant.domain.service.LocationService
import com.google.android.gms.location.FusedLocationProviderClient
import com.google.android.gms.location.LocationCallback
import com.google.android.gms.location.LocationRequest
import com.google.android.gms.location.LocationResult
import com.google.android.gms.location.LocationServices
import com.google.android.gms.location.Priority
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.suspendCancellableCoroutine
import java.util.Date
import java.util.UUID
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.coroutines.resume

private val Context.dataStore by preferencesDataStore(name = "saved_locations")
private val SAVED_LOCATIONS_KEY = stringPreferencesKey("saved_locations")

@Singleton
class LocationServiceImpl @Inject constructor(
    @ApplicationContext private val context: Context,
    private val locationPermissionManager: LocationPermissionManager
) : LocationService {

    private val fusedLocationClient: FusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(context)
    private val _currentLocation = MutableStateFlow<Location?>(null)
    
    // Shared preferences for location caching
    private val sharedPreferences: SharedPreferences = context.getSharedPreferences(
        "location_cache", Context.MODE_PRIVATE
    )
    
    // Battery optimization settings
    private var isMoving = false
    private var lastSignificantMovement = Date()
    private val significantMovementThresholdMs = 5 * 60 * 1000 // 5 minutes
    
    // Location accuracy modes
    private var currentAccuracyMode = AccuracyMode.BALANCED
    
    // Gson for JSON serialization
    private val gson = Gson()
    
    // Keys for cached location
    private val CACHED_LOCATION_KEY = "cached_location"
    private val CACHED_LOCATION_TIME_KEY = "cached_location_time"
    
    enum class AccuracyMode {
        HIGH_ACCURACY,
        BALANCED,
        LOW_POWER
    }
    
    // Create location requests for different accuracy modes
    private val highAccuracyRequest = LocationRequest.Builder(Priority.PRIORITY_HIGH_ACCURACY, TimeUnit.SECONDS.toMillis(5))
        .setWaitForAccurateLocation(true)
        .setMinUpdateIntervalMillis(TimeUnit.SECONDS.toMillis(2))
        .setMaxUpdateDelayMillis(TimeUnit.SECONDS.toMillis(10))
        .build()
    
    private val balancedRequest = LocationRequest.Builder(Priority.PRIORITY_BALANCED_POWER_ACCURACY, TimeUnit.SECONDS.toMillis(15))
        .setWaitForAccurateLocation(false)
        .setMinUpdateIntervalMillis(TimeUnit.SECONDS.toMillis(10))
        .setMaxUpdateDelayMillis(TimeUnit.SECONDS.toMillis(30))
        .build()
    
    private val lowPowerRequest = LocationRequest.Builder(Priority.PRIORITY_LOW_POWER, TimeUnit.MINUTES.toMillis(1))
        .setWaitForAccurateLocation(false)
        .setMinUpdateIntervalMillis(TimeUnit.SECONDS.toMillis(30))
        .setMaxUpdateDelayMillis(TimeUnit.MINUTES.toMillis(2))
        .build()
    
    private val locationCallback = object : LocationCallback() {
        override fun onLocationResult(result: LocationResult) {
            result.lastLocation?.let { location ->
                handleNewLocation(location)
            }
        }
    }
    
    init {
        // Load cached location on startup
        loadCachedLocation()?.let {
            _currentLocation.value = it
        }
        
        if (locationPermissionManager.hasLocationPermission()) {
            startLocationUpdates()
        }
    }
    
    override val currentLocation: Flow<Location?> = _currentLocation
    
    override val permissionStatus: Flow<LocationPermissionStatus> = locationPermissionManager.permissionStatus
    
    override suspend fun requestLocationPermission() {
        locationPermissionManager.requestLocationPermission()
        if (locationPermissionManager.hasLocationPermission()) {
            startLocationUpdates()
        }
    }
    
    @SuppressLint("MissingPermission")
    override suspend fun getCurrentLocation(): Result<Location?> = runCatching {
        if (!locationPermissionManager.hasLocationPermission()) {
            return@runCatching null
        }
        
        // Check if we have a recent location
        val cachedLocation = _currentLocation.value
        if (cachedLocation != null) {
            // If the location is recent (less than 1 minute old), use it
            if (SystemClock.elapsedRealtimeNanos() - cachedLocation.elapsedRealtimeNanos < TimeUnit.MINUTES.toNanos(1)) {
                return@runCatching cachedLocation
            }
        }
        
        // Check if we have a stored location that's still valid
        loadCachedLocation()?.let { storedLocation ->
            // Use the stored location while requesting a fresh one in the background
            requestFreshLocation()
            return@runCatching storedLocation
        }
        
        // Request a fresh location
        return@runCatching suspendCancellableCoroutine { continuation ->
            // Temporarily switch to high accuracy
            val previousMode = currentAccuracyMode
            setAccuracyMode(AccuracyMode.HIGH_ACCURACY)
            
            fusedLocationClient.lastLocation
                .addOnSuccessListener { location ->
                    if (location != null) {
                        handleNewLocation(location)
                        continuation.resume(location)
                    } else {
                        // If last location is null, request a single update
                        val singleLocationRequest = LocationRequest.Builder(Priority.PRIORITY_HIGH_ACCURACY, 10000)
                            .setMaxUpdates(1)
                            .build()
                        
                        val singleLocationCallback = object : LocationCallback() {
                            override fun onLocationResult(result: LocationResult) {
                                result.lastLocation?.let { newLocation ->
                                    handleNewLocation(newLocation)
                                    continuation.resume(newLocation)
                                } ?: continuation.resume(null)
                                fusedLocationClient.removeLocationUpdates(this)
                                
                                // Reset accuracy mode
                                setAccuracyMode(previousMode)
                            }
                        }
                        
                        fusedLocationClient.requestLocationUpdates(
                            singleLocationRequest,
                            singleLocationCallback,
                            Looper.getMainLooper()
                        )
                        
                        continuation.invokeOnCancellation {
                            fusedLocationClient.removeLocationUpdates(singleLocationCallback)
                            setAccuracyMode(previousMode)
                        }
                    }
                }
                .addOnFailureListener { exception ->
                    continuation.resume(null)
                    setAccuracyMode(previousMode)
                }
        }
    }
    
    override suspend fun searchLocations(query: String): Result<List<DomainLocation>> = runCatching {
        val geocoder = Geocoder(context)
        val addresses = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            suspendCancellableCoroutine { continuation ->
                geocoder.getFromLocationName(query, 10) { addresses ->
                    continuation.resume(addresses ?: emptyList())
                }
            }
        } else {
            @Suppress("DEPRECATION")
            geocoder.getFromLocationName(query, 10) ?: emptyList()
        }
        
        return@runCatching addresses.map { address ->
            addressToDomainLocation(address)
        }
    }
    
    private val locationDao: LocationDao
    
    init {
        val db = Room.databaseBuilder(
            context,
            AppDatabase::class.java,
            "salmon_trolling_assistant_db"
        ).build()
        locationDao = db.locationDao()
    }
    
    override fun getSavedLocations(): Flow<List<DomainLocation>> = 
        locationDao.getAllLocations().map { entities ->
            entities.map { it.toDomainModel() }
        }
    
    override suspend fun saveLocation(location: DomainLocation): Result<Boolean> = runCatching {
        val locationToSave = location.copy(isSaved = true)
        locationDao.insertLocation(LocationEntity.fromDomainModel(locationToSave))
        true
    }
    
    override suspend fun deleteLocation(locationId: String): Result<Boolean> = runCatching {
        locationDao.deleteLocationById(locationId)
        true
    }
    
    @SuppressLint("MissingPermission")
    private fun startLocationUpdates() {
        if (!locationPermissionManager.hasLocationPermission()) {
            return
        }
        
        // Start with balanced accuracy
        setAccuracyMode(AccuracyMode.BALANCED)
    }
    
    @SuppressLint("MissingPermission")
    private fun setAccuracyMode(mode: AccuracyMode) {
        if (!locationPermissionManager.hasLocationPermission()) {
            return
        }
        
        // Remove any existing callbacks
        fusedLocationClient.removeLocationUpdates(locationCallback)
        
        currentAccuracyMode = mode
        
        // Select the appropriate request based on mode
        val request = when (mode) {
            AccuracyMode.HIGH_ACCURACY -> highAccuracyRequest
            AccuracyMode.BALANCED -> balancedRequest
            AccuracyMode.LOW_POWER -> lowPowerRequest
        }
        
        // Request location updates with the new settings
        fusedLocationClient.requestLocationUpdates(
            request,
            locationCallback,
            Looper.getMainLooper()
        )
        
        // If in low power mode, schedule to stop updates after a while
        if (mode == AccuracyMode.LOW_POWER) {
            android.os.Handler(Looper.getMainLooper()).postDelayed({
                if (currentAccuracyMode == AccuracyMode.LOW_POWER) {
                    fusedLocationClient.removeLocationUpdates(locationCallback)
                }
            }, TimeUnit.MINUTES.toMillis(5))
        }
    }
    
    @SuppressLint("MissingPermission")
    private fun requestFreshLocation() {
        if (!locationPermissionManager.hasLocationPermission()) {
            return
        }
        
        // Save current mode
        val previousMode = currentAccuracyMode
        
        // Switch to high accuracy temporarily
        setAccuracyMode(AccuracyMode.HIGH_ACCURACY)
        
        // Reset after a short time
        android.os.Handler(Looper.getMainLooper()).postDelayed({
            setAccuracyMode(previousMode)
        }, 10000) // 10 seconds
    }
    
    private fun handleNewLocation(location: Location) {
        val previousLocation = _currentLocation.value
        
        // Update the current location
        _currentLocation.value = location
        
        // Cache the location
        cacheLocation(location)
        
        // Check if this is a significant movement
        if (previousLocation != null) {
            val distance = location.distanceTo(previousLocation)
            val timeDiff = location.time - previousLocation.time
            
            // If moved more than 100 meters in less than 5 minutes, consider as moving
            if (distance > 100 && timeDiff < significantMovementThresholdMs) {
                isMoving = true
                lastSignificantMovement = Date()
                
                // If moving, increase accuracy if not already high
                if (currentAccuracyMode != AccuracyMode.HIGH_ACCURACY) {
                    setAccuracyMode(AccuracyMode.HIGH_ACCURACY)
                }
            } else {
                // Check if we haven't moved significantly for a while
                if (System.currentTimeMillis() - lastSignificantMovement.time > significantMovementThresholdMs) {
                    isMoving = false
                    
                    // If not moving, decrease accuracy to save battery
                    if (currentAccuracyMode == AccuracyMode.HIGH_ACCURACY) {
                        setAccuracyMode(AccuracyMode.BALANCED)
                    }
                }
            }
        } else {
            // First location update
            lastSignificantMovement = Date()
        }
    }
    
    private fun addressToDomainLocation(address: Address): DomainLocation {
        return DomainLocation(
            id = UUID.randomUUID().toString(),
            name = address.getAddressLine(0) ?: "Unknown Location",
            latitude = address.latitude,
            longitude = address.longitude,
            isSaved = false,
            notes = null
        )
    }
    
    // Location caching methods
    private fun cacheLocation(location: Location) {
        try {
            // Create a serializable location object
            val locationMap = mapOf(
                "latitude" to location.latitude,
                "longitude" to location.longitude,
                "accuracy" to location.accuracy,
                "altitude" to location.altitude,
                "speed" to location.speed,
                "bearing" to location.bearing,
                "time" to location.time
            )
            
            val locationJson = gson.toJson(locationMap)
            
            // Save to shared preferences
            sharedPreferences.edit()
                .putString(CACHED_LOCATION_KEY, locationJson)
                .putLong(CACHED_LOCATION_TIME_KEY, System.currentTimeMillis())
                .apply()
        } catch (e: Exception) {
            // Log error but don't crash
            e.printStackTrace()
        }
    }
    
    private fun loadCachedLocation(): Location? {
        try {
            val locationJson = sharedPreferences.getString(CACHED_LOCATION_KEY, null) ?: return null
            val timestamp = sharedPreferences.getLong(CACHED_LOCATION_TIME_KEY, 0)
            
            // Check if cache is too old (more than 1 hour)
            if (System.currentTimeMillis() - timestamp > TimeUnit.HOURS.toMillis(1)) {
                return null
            }
            
            // Parse the location
            val locationMap = gson.fromJson<Map<String, Any>>(
                locationJson,
                object : TypeToken<Map<String, Any>>() {}.type
            )
            
            // Create a new Location object
            val location = Location("cached")
            location.latitude = (locationMap["latitude"] as Number).toDouble()
            location.longitude = (locationMap["longitude"] as Number).toDouble()
            location.accuracy = (locationMap["accuracy"] as Number).toFloat()
            location.altitude = (locationMap["altitude"] as Number).toDouble()
            location.speed = (locationMap["speed"] as Number).toFloat()
            location.bearing = (locationMap["bearing"] as Number).toFloat()
            location.time = (locationMap["time"] as Number).toLong()
            
            return location
        } catch (e: Exception) {
            // Log error but don't crash
            e.printStackTrace()
            return null
        }
    }
    
    private fun parseLocationsFromJson(json: String): List<DomainLocation> {
        return try {
            gson.fromJson(json, object : TypeToken<List<DomainLocation>>() {}.type) ?: emptyList()
        } catch (e: Exception) {
            emptyList()
        }
    }
    
    private fun locationsToJson(locations: List<DomainLocation>): String {
        return try {
            gson.toJson(locations)
        } catch (e: Exception) {
            "[]"
        }
    }
}